data/synth_timing.xdc

```diff
@@ -4,29 +4,860 @@
+#
+# *1: Data Valid minimum period = the lesser of:
+#   (CK half-period min - output max-dly to valid + output max-dly to valid) or
+#   (CK half-period min - output min-dly to valid + output min-dly to valid)
+#
+# TODO: add 'real' constraints here and remove false_paths so we know
+#       if something has not been instantiated/inferred correctly.
+
+### Clock Groups and Clock False Paths ###
+# JTAG tck is completely asynchronous to FPGA mainClk and derivatives
+set_clock_groups -asynchronous -group tck -group $mainClk_and_generated
+
+
+#### Timing Exceptions Section ####
+
+### False Paths ###
+## Reset - asynchronous on-board button or external source.
+# Set as false path and use deassert-synchroniser.
+set_false_path -from [get_ports nrst]
+
+## General purpose LEDs - human-scale asynchronous
+set_false_path -to [get_ports {usrLed[*]}]
+## CHERI error LEDs - human-scale asynchronous
+set_false_path -to [get_ports {cheriErr[*]}]
+## Status LEDs - human-scale asynchronous
+set_false_path -to [get_ports {led_legacy led_cheri led_halted led_bootok}]
+## Switch and button input - human-scale asynchronous
+set_false_path -from [get_ports {usrSw[*]}]
+set_false_path -from [get_ports {navSw[*]}]
+set_false_path -from [get_ports {selSw[*]}]
+
+## USRUSB interface - asynchronous vbus detection
+set_false_path -from [get_ports usrusb_vbusdetect]
+
+## LCD display
+# Reset must be held for 9+ us.
+# Backlight signal drives gate of discrete transistor that powers backlights.
+set_false_path -to [get_ports {lcd_rst lcd_backlight}] ;# vbus detection - async
+
+## Ethernet MAC - asynchronous reset
+set_false_path -to [get_ports ethmac_rst]
+
+## HyperRAM
+# Constraints (and RTL) are adapted from OpenHBMC:
+# https://github.com/OVGN/OpenHBMC/blob/master/OpenHBMC/constrs/OpenHBMC.xdc
+# The RTL instantiates I/O primitives itself and places flop in IOBs,
+# reducing the need for I/O timing constraints.
+#
+# TODO: replace these with 'real' constraints so we know if something has
+#       not been instantiated/inferred correctly.
+#
+# Set output false path, timings are met by design
 set_false_path -to [get_ports hyperram_ckp]
 set_false_path -to [get_ports hyperram_ckn]
 set_false_path -to [get_ports hyperram_rwds]
-set_false_path -to [get_ports hyperram_dq[*]]
-
+set_false_path -to [get_ports {hyperram_dq[*]}]
 # set input false path. dq[*] and rwds are supposed to
 # be fully asynchronous for the data recovery logic
 set_false_path -from [get_ports hyperram_rwds]
-set_false_path -from [get_ports hyperram_dq[*]]
-
+set_false_path -from [get_ports {hyperram_dq[*]}]
 # False path for 'hb_cs_n' and 'hb_reset_n'
 set_false_path -to [get_ports hyperram_cs]
 set_false_path -to [get_ports hyperram_nrst]
+
+### Max Delay / Min Delay ###
+## Ethernet MAC
+# Asynchronous(?) interrupt.
+# Use max_delay rather than false_path to keep the delay within reason.
+set_max_delay 30 -datapath_only -from [get_ports ethmac_intr]
+
+### Multicycle Paths ###
```

Comment:
Were any of these multi-cycle paths necessary to close timing? Perhaps we can get some better implementation results by giving the optimizer a easier job but I am very wary of including multi-cycle paths. In particular perhaps something is indeed multi-cycle right now but after some RTL change it isn't and these don't get appropriately adjusted.

So I'd probably drop them all unless there's specific cases you've found they're required. Will also be more complicated once pinmux is in there.

Reply:
The UARTs definitely need some sort of timing exception because of the combinatorial line-loopback path from RX to TX. I used a multicycle path instead of a false path because if doesn't disable as many design checks.

I'll run a trial to check what happens when I remove the multipath constraints from the non-UART interfaces. Might do a rebase to include pinmux first to avoid later surprises.

Reply:
Thanks for the thorough analysis @elliotb-lowrisc hadn't realised we were looking so tight on the I/O timing. Given this analysis I concur with you the multi-cycle paths are worth keeping, though clearly we need to exercise caution around any RTL that involves them. I wonder if we can do something in simulation with assertions that fire if the relevant signals change more frequently than every 2 clocks?

Reply:
Oh and we are hitting utilization limits which may explain the routing and timing issues. Various things in flight to bring that down significantly.